Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+                                        thisParent->sgroups.end(), this));
+
+    if (thisParent == actualParent) {
+      break;
+    }
+  }
+
+  delete this;
+}
+
+void MarvinMolBase::prepSgroupsForRDKit() {
+  // this routine recursively fixes the hierarchy of sgroups - some may NOT
+  // actaully belong underneath their parent and can be promnoted to the
+  // grandparent
+
+  // first fix all the children
+
+  std::vector<std::string> sgroupsMolIdsDone;
+  for (bool allDone = false;
+       !allDone;)    // until all at this level are done - but fixing one child
+                     // can add sgroups to this level
+  {
+    allDone = true;  // unitl it is not true in the loop below
+    for (auto childSgroup : this->sgroups) {
+      if (!boost::algorithm::contains(
+              sgroupsMolIdsDone, std::vector<std::string>{childSgroup->molID},
+              molIDInSgroups)) {
+        sgroupsMolIdsDone.push_back(childSgroup->molID);
+        childSgroup->prepSgroupsForRDKit();
+        allDone = false;
+        break;  // have to start the loop over - we might have changed the
+                // vector
+      }
+    }
+  }
+  // if this is top level, quit
+
+  if (this->parent == nullptr) {
+    return;
+  }
+
+  // check to see if we can move this sgroup up a level. We can if the parent
+  // is not the root of the tree, and its parent is passive (it does not
+  // change the atoms of the parent when it is processed or expanded, or if
+  // all of its atoms do not belong to the the non-passive parent
+
+  std::string parentRole = this->parent->role();
+  std::string childRole = this->role();
+
+  if (parentRole != "MarvinMol") {
+    // check this one
+
+    if (this->parent->isPassiveRoleForExpansion()) {
+      // move the child up
+
+      promoteChild(this);
+      return;
+    }
+
+    // check to see if all of its atomRefs of the child are in the parent.
+    //  there are 3 possibilities:
+    //  1) all atoms are in the parent - do NOT move the child - it belongs
+    //  under the parent 2) NO atoms are in the parent - move the child up 3)
+    //  SOME atoms are in the parent - throw an error
+
+    auto result = this->isSgroupInSetOfAtoms(this->parent->atoms);
+    switch (result) {
+      case SgroupNotInAtomSet:
+        promoteChild(this);
+        return;
+
+      case SgroupInAtomSet:
+        // do noting here
+        break;
+
+      case SgroupBothInAndNotInAtomSet:
+        throw FileParseException(
+            "Child sGroup has atoms both in the parent and NOT in the parent");
+    }
+  }
+
+  // made to here, so the child belongs in the parent.
+
+  if (childRole == "SuperatomSgroup") {
+    ((MarvinSuperatomSgroup *)this)->convertFromOneSuperAtom();
+  } else if (childRole == "MulticenterSgroup") {
+    ((MarvinMulticenterSgroup *)this)->processOneMulticenterSgroup();
+  } else if (childRole == "MultipleSgroup") {
+    ((MarvinMultipleSgroup *)this)->expandOneMultipleSgroup();
+  }
```

Comment:
what if `childRole` is none of these values?

Reply:
this code no longer exists